src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -656,70 +653,204 @@ IEnumerator IEnumerable.GetEnumerator()
             return GetEnumerator();
         }
 
-        public IEnumerable<KeyValuePair<string, ModelState>> FindKeysWithPrefix(string prefix)
+        public static bool StartsWithPrefix(string prefix, string key)
         {
             if (prefix == null)
             {
                 throw new ArgumentNullException(nameof(prefix));
             }
 
-            ModelState exactMatchValue;
-            if (_innerDictionary.TryGetValue(prefix, out exactMatchValue))
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (StringComparer.OrdinalIgnoreCase.Equals(key, prefix))
+            {
+                return true;
+            }
+
+            if (key.Length <= prefix.Length)
             {
-                yield return new KeyValuePair<string, ModelState>(prefix, exactMatchValue);
+                return false;
             }
 
-            foreach (var entry in _innerDictionary)
+            if (!key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
             {
-                var key = entry.Key;
+                if (key.StartsWith("[", StringComparison.OrdinalIgnoreCase))
+                {
+                    var subKey = key.Substring(key.IndexOf('.') + 1);
```

Comment:
This could obviously be optimized further, this isn't the part I'm going after. Will wait for some more data to motivate further changes.